### PR TITLE
CORE-5011 Add support for soak test Java thread/heap/OOM dumps and YourKit profiling

### PR DESCRIPTION
**Pull request by oscardalmau-r3**
Tickets involved:
https://r3-cev.atlassian.net/browse/CORE-5011
https://r3-cev.atlassian.net/browse/CORE-4990
https://r3-cev.atlassian.net/browse/CORE-4991
https://r3-cev.atlassian.net/browse/CORE-4992
https://r3-cev.atlassian.net/browse/CORE-5078

- Add pod **security context** function in helpers:
    - Required to grant access to the pod to create the log files for the java dumps and application profiling.
    - When profiling is enabled on the workers or _dumpHostPath_ is defined, the pod user will be root instead of 1000.

- Update **volumeMounts** and **volumes** functions:
    -  When _dumpHostPath_ is defined, it triggers the setting up of a _/dumps_ mount path on all of the workers and a hostPath volume on the desired path.

- Update worker templates files to use the functions mentioned above and to expose the profiling port when enabled.

- Update **workerEnv** function:
    -  Include additional JAVA TOOL OPTIONS for enabling JVM dump on OOM.
    - Update JAVA TOOL OPTIONS _agentpath_ value to include _dir_ and _logDir_ flags when _dumpHostPath_ is defined,

### CODE REVIEW

File: charts/corda/templates/_helpers.tpl [hunk 1]
```diff
@@ -163,9 +175,16 @@ Worker environment variables
...
-    {{- if ( get .Values.workers .worker ).profiling.enabled }}
+    {{- if and (( get .Values.workers .worker ).profiling.enabled ) ( not .Values.dumpHostPath )  }}
       -agentpath:/opt/override/libyjpagent.so=exceptions=disable,port=10045,listen=all
     {{- end -}}
+    {{- if and (( get .Values.workers .worker ).profiling.enabled ) ( .Values.dumpHostPath )  }}
+      -agentpath:/opt/override/libyjpagent.so=exceptions=disable,port=10045,listen=all,dir=/dumps/profile/snapshots,logdir=/dumps/profile/logs
+    {{- end -}}
+    {{- if .Values.heapDumpOnOutOfMemoryError }}
+      -XX:+HeapDumpOnOutOfMemoryError
+      -XX:HeapDumpPath=/dumps/heap
```
Comment (oscardalmau-r3) [inline charts/corda/templates/_helpers.tpl:183]: @davidcurrie should the HeapDumpPath for OOM and "normal" JVM heap dump be the same? or should this one be "/dumps/oom-heap"?

Comment (davidcurrie) [reply charts/corda/templates/_helpers.tpl:183]: I think `/heap` is fine for both.

File: charts/corda/templates/_helpers.tpl [hunk 2]
```diff
@@ -99,6 +99,18 @@ Worker image
 "{{ ( get .Values.workers .worker ).image.registry | default .Values.image.registry }}/{{ ( get .Values.workers .worker ).image.repository }}:{{ ( get .Values.workers .worker ).image.tag | default .Values.image.tag | default .Chart.AppVersion }}"
 {{- end }}
 
+{{/*
+Worker security context
+*/}}
+{{- define "corda.workerSecurityContext" -}}
+{{- if and ( not .Values.dumpHostPath ) ( not ( get .Values.workers .worker ).profiling.enabled ) }}
```
Comment (davidcurrie) [inline charts/corda/templates/_helpers.tpl:106]: I think we only need to check `dumpHostPath` here. If profiling is enabled but it's only writing inside the container then we can stick with the default security context.

File: charts/corda/templates/_helpers.tpl [hunk 3]
```diff
@@ -163,9 +175,16 @@ Worker environment variables
   value: {{- if ( get .Values.workers .worker ).debug.enabled }}
       -agentlib:jdwp=transport=dt_socket,server=y,address=5005,suspend={{ if ( get .Values.workers .worker ).debug.suspend }}y{{ else }}n{{ end }}
     {{- end -}}
-    {{- if ( get .Values.workers .worker ).profiling.enabled }}
+    {{- if and (( get .Values.workers .worker ).profiling.enabled ) ( not .Values.dumpHostPath )  }}
       -agentpath:/opt/override/libyjpagent.so=exceptions=disable,port=10045,listen=all
     {{- end -}}
+    {{- if and (( get .Values.workers .worker ).profiling.enabled ) ( .Values.dumpHostPath )  }}
+      -agentpath:/opt/override/libyjpagent.so=exceptions=disable,port=10045,listen=all,dir=/dumps/profile/snapshots,logdir=/dumps/profile/logs
+    {{- end -}}
```
Comment (davidcurrie) [inline charts/corda/templates/_helpers.tpl]: I'd always just set `dir` and `logdir` to use those directories under `/dumps`. If there's no mount on `/dumps` that's fine - we'll just write to that location inside the container.

File: charts/corda/values.yaml [hunk 4]
```diff
@@ -35,6 +35,12 @@ logging:
   # -- log level; one of "all", "trace", "debug", "info", "warn" (the default), "error", "fatal", or "off"
   level: "warn"
 
+# Dump host path configuration
```
Comment (davidcurrie) [inline charts/corda/values.yaml]:
```suggestion
# -- Path on Kubernetes hosts to mount on Corda workers for collecting dumps
```

File: charts/corda/values.yaml [hunk 5]
```diff
@@ -35,6 +35,12 @@ logging:
   # -- log level; one of "all", "trace", "debug", "info", "warn" (the default), "error", "fatal", or "off"
   level: "warn"
 
+# Dump host path configuration
+dumpHostPath: ""
+
+# Enables capturing JVM heap memory dump from unexpectedly terminated pods
```
Comment (davidcurrie) [inline charts/corda/values.yaml]:
```suggestion
# -- Enables capturing JVM heap dumps from Corda workers on an OutOfMemory error

```

File: gradle.properties [hunk 6]
```diff
@@ -138,7 +138,7 @@ postgresUser=
 postgresPassword=
 
 # Profiling support
-enableProfiling=false
+enableProfiling=true
```
Comment (davidcurrie) [inline gradle.properties]: Don't forget... 😄

File: charts/corda/templates/_helpers.tpl (hunk 2 above)
Comment (oscardalmau-r3) [reply charts/corda/templates/_helpers.tpl:106]: We need this in both cases. When profiling is enabled and we set dir and logdir in the agentpath, it fails if the pod user don't have access to create directories within the container.

File: charts/corda/templates/_helpers.tpl (hunk 3 above)
Comment (oscardalmau-r3) [reply charts/corda/templates/_helpers.tpl]: Makes sense

File: gradle.properties (hunk 6 above)
Comment (oscardalmau-r3) [reply gradle.properties]: Will remove this before merging, want to test a soak test deployment with this enabled first.

File: charts/corda/templates/_helpers.tpl (hunk 2 above)
Comment (davidcurrie) [reply charts/corda/templates/_helpers.tpl:106]: That's something we fix but not under this PR.

Comment (oscardalmau-r3) [reply charts/corda/templates/_helpers.tpl:106]: Yeah agree, we tried using _privileged_ but didn't work. We probably need to take a look at the Dockerfile to fix this and do it the proper way.

File: charts/corda/values.yaml [hunk 7]
```diff
@@ -35,6 +35,13 @@ logging:
   # -- log level; one of "all", "trace", "debug", "info", "warn" (the default), "error", "fatal", or "off"
   level: "warn"
 
+# -- Path on Kubernetes hosts to mount on Corda workers for collecting dumps
+dumpHostPath: ""
+
+# -- Enables capturing JVM heap dumps from Corda workers on an OutOfMemory error
+
```
Comment (davidcurrie) [inline charts/corda/values.yaml]: My fault... the GitHub suggestions always try to add an extra newline...
```suggestion
```

**Review by davidcurrie — CHANGES_REQUESTED** [review]
One tiny change otherwise LGTM

**Review by davidcurrie — DISMISSED** [review]
👍 LGTM

File: charts/corda/templates/_helpers.tpl [hunk 8]
```diff
@@ -69,6 +69,13 @@ Worker name
 "{{ include "corda.fullname" . }}-{{ .worker | kebabcase | replace "p-2p" "p2p" }}-worker"
 {{- end }}
 
+{{/*
+Worker name with no quotes
+*/}}
+{{- define "corda.workerNameNoQuotes" -}}
+{{ include "corda.fullname" . }}-{{ .worker | kebabcase | replace "p-2p" "p2p" }}-worker
+{{- end }}
+
```
Comment (davidcurrie) [inline charts/corda/templates/_helpers.tpl]: I'd just take the quotes out of `corda.workerName` and change the existing places it is used to `{{ include "corda.workerName" . | quote }}`

File: gradle.properties [hunk 9]
```diff
@@ -50,6 +50,7 @@ felixSecurityVersion=2.8.3
 #  Check with one of the group leads before changing.
 guavaVersion=30.1.1-jre
 hibernateVersion = 5.6.10.Final
+      oerviceAccountName: privileged
```
Comment (davidcurrie) [inline gradle.properties]: don't think this should be here!

**Review by davidcurrie — DISMISSED** [review]
👍

**Review by davidcurrie — APPROVED** [review]

**Pull request merged**